### PR TITLE
Force `npm install`

### PR DESCRIPTION
**Pull request by 0x2b3bfa0**
@RCdeWit has [found](https://iterativeai.slack.com/archives/CFB1AJV8U/p1649687067838479?thread_ts=1649685991.288849&cid=CFB1AJV8U) an error when using this `iterative/setup-cml` action in a self–hosted runner that, of course, already had `cml` installed.

@DavidGOrtega edits
~~This pull request adds the `-f` option to _“overwrite files recklessly [sic]”_ and workaround this issue.~~

 - Checks if existing version coincides with desired version and if so do nothing
 - Add `force` parameter to force the installation
 - Logs the install output to be able to understand whats going on 

The interesting/tricky thing is that `uninstall` then `install` does not always fail. However `force` is the only way of guarantee the complete installation.
For the majority of the cases that this could happen (runner image and installed latest cml version) this code may suffice perfectly

### CODE REVIEW

**Review by DavidGOrtega — DISMISSED** [review]
👍

**Review by casperdcl — DISMISSED** [review]

**Comment by casperdcl** [comment]
- isn't the action meant to "uninstall the previous CML" first though?
- also windows tests failing

**Comment by DavidGOrtega** [comment]
> isn't the action meant to "uninstall the previous CML" first though?

Exactly! The prior version should be removed first

**Comment by DavidGOrtega** [comment]
Changed p0 because seems to be needed for the release

**Comment by 0x2b3bfa0** [comment]
> _isn't the action meant to "uninstall the previous CML" first though?_

The previous CML is the binary we use to launch the runner. Not sure we actually want to uninstall that, or even overwrite it for that effect. 🤔

**Comment by 0x2b3bfa0** [comment]
> _also windows tests failing_

(Please kill me)

**Comment by DavidGOrtega** [comment]
> The previous CML is the binary we use to launch the runner. Not sure we actually want to uninstall that, or even overwrite it for that effect. 🤔

😆 true!

But it works anyway

**Comment by 0x2b3bfa0** [comment]
Yes, it works anyway or, at least, it should.

**Comment by 0x2b3bfa0** [comment]
⚠️ **On a second thought, overwriting the executable may crash the program**

When you [`unlink(1)`](https://man7.org/linux/man-pages/man2/unlink.2.html) an executable while it's running, the file doesn't actually get deleted until the executable exits and the [file descriptor](https://en.wikipedia.org/wiki/File_descriptor) reference count is zero. It has no effect on the running process.

When you **overwrite** an executable while it's running it may crash in very creative ways, because it's not loaded in memory at once; pages are loaded on demand.

**Comment by 0x2b3bfa0** [comment]
@DavidGOrtega, TL;DR, it's fine to replace it when it's running, but we need to make sure that it's doing `rm old && cp new old` (delete & create) instead of `cat new > old` (overwrite)

**Comment by casperdcl** [comment]
would #57 be a better fix?

**Comment by 0x2b3bfa0** [comment]
Yes, definitely, but we need to build Windows binaries, and https://github.com/iterative/setup-cml/pull/60#issuecomment-1097120247 still applies.

**Comment by DavidGOrtega** [comment]
@0x2b3bfa0 I knew! Yesterday I was trying to understand what is doing `npm uninstall`.

😉 
```yaml
- name: test cml
      run: |
        node --version
        cml --version
        # which cml
        # rm /usr/bin/cml
        npm uninstall -g @dvcorg/cml
        cml --version || echo 'no cml'
```

**Comment by DavidGOrtega** [comment]
> When you overwrite an executable while it's running it may crash in very creative ways, because it's not loaded in memory at once; pages are loaded on demand.

My rationale is that in case of a node app its fine while we do not do it against node itself. Our code its just interpreted by the v8 engine, loaded in memory by the VM

**Comment by 0x2b3bfa0** [comment]
> _My rationale is that in case of a node app its fine while we do not do it against node itself._

The `pkg` binaries are “node itself” plus our code, so yes, it's equally risky

> _Our code its just interpreted by the v8 engine, loaded in memory by the VM_

Not sure... does V8 load _all_ the code at once, or waits until it's needed to read the file? 🤔

**Comment by DavidGOrtega** [comment]
> The pkg binaries are “node itself” plus our code, so yes, it's equally risky

I think I have not explained my self. Using the bin is more dangerous, the `npm install` its more secure because its the V8 VM loading our js

> Not sure... does V8 load all the code at once, or waits until it's needed to read the file?

Everything unless we do the `require` inside non evaluated code. But we do the require in every single module on top.
But I might be wrong

**Comment by 0x2b3bfa0** [comment]
🤦🏼 Makes sense now. I thought you were talking about binaries.

**Comment by 0x2b3bfa0** [comment]
Still, replacing binaries safely is straightforward (e.g. `rm` and `cp` as stated above) while installing a Node.js package might be a bit less stable. 🤔

**Comment by DavidGOrtega** [comment]
At current stage the workflow hangs forever in macos, something that does not happen installing the action.

```yaml
jobs:
 cml:
    runs-on: [macos-latest]
    env:
      REPO_TOKEN: ${{ secrets.REPO_TOKEN }}
    steps:
      - uses: actions/checkout@v2
      - uses: iterative/setup-cml@v1
```

**Comment by 0x2b3bfa0** [comment]
@DavidGOrtega, https://github.com/iterative/cml/pull/983 solves all your hanging issues in a rather drastic way. 🐉

**Comment by DavidGOrtega** [comment]
Unfortunately does not totally fix it for previous version of CML and we could end up with the same error but I have remembered this issue and I can pipe a dummy text as we do in formal CML tests.

I had to remove publish tests: In windows they fail. https://github.com/iterative/cml/pull/901 seems not to have fixed it?

**Comment by DavidGOrtega** [comment]
@0x2b3bfa0

**Comment by DavidGOrtega** [comment]
Also tests in container fails until https://github.com/iterative/cml/pull/974 is effective

**Comment by DavidGOrtega** [comment]
@iterative/cml ready to review 🙏

**Comment by 0x2b3bfa0** [comment]
> _I had to remove publish tests: In windows they fail. https://github.com/iterative/cml/pull/901 seems not to have fixed it?_

Windows tests are broken because an arcane issue involving the drive letter of the current working directory not being the same as the drive letter of the [`@npcz/magic`](https://github.com/npcz/magic) installation path. 🎩 [Magic.](http://www.catb.org/jargon/html/magic-story.html)
* https://github.com/iterative/setup-cml/issues/56
* https://github.com/iterative/cml/pull/865

File: src/utils.js [hunk 1]
```diff
@@ -27,19 +29,29 @@ const setupCml = async opts => {
     }
   }
 
-  await exec(`${sudoPath} npm config set user 0`);
+  try {
+    const cmlVer = await exec('echo none | cml --version');
```
Comment (0x2b3bfa0) [inline src/utils.js]: Arcane code golf alternatives 😄 
```suggestion
    const cmlVer = await exec(':|cml --version');
```
```suggestion
    const cmlVer = await exec('cml --version<&-');
```

**Comment by 0x2b3bfa0** [comment]
I can't approve “my own” pull request, by the way. 🙃

**Comment by 0x2b3bfa0** [comment]
Attracting reviewers from @iterative/cml...

File: src/utils.js (hunk 1 above)
Comment (dacbd) [reply src/utils.js]: might include a comment [link](https://www.gnu.org/software/bash/manual/bash.html#Redirections) for the others who may try their hand at contributing and arent bash ninjas

**Review by dacbd — CHANGES_REQUESTED** [review]
Revert https://github.com/iterative/setup-cml/pull/60/commits/3beb66c570ecfbc6e33e6605b8127c8039f8d10e or explain the apparent lock-file rollback?

**Comment by DavidGOrtega** [comment]
@dacbd ready for review 🙏

**Review by dacbd — APPROVED** [review]
https://github.com/iterative/setup-cml/issues/65, see: https://github.com/iterative/setup-cml/runs/6165246456?check_suite_focus=true#step:3:8

**Comment by 0x2b3bfa0** [comment]
@RCdeWit, [this](https://iterativeai.slack.com/archives/CFB1AJV8U/p1649687067838479?thread_ts=1649685991.288849&cid=CFB1AJV8U) should be fixed now.

**Pull request merged**